lookup/lookup.js

```diff
@@ -59,17 +61,44 @@ async function main (params) {
   };
 }
 
-async function lookup (args) {
+function lookup (args) {
   const agreements = args.agreements;
-  let usernames = [];
-  await Promise.all(agreements.map(async function (agreement) {
-    const agreementUsers = await lookupSingleAgreement(args, agreement);
-    usernames = Array.from(new Set(usernames.concat(agreementUsers)));
-  }));
-  return usernames;
+  const username = args.username.toLowerCase();
+  const promiseMatrix = agreements.map(agreement => lookupAndParseAgreement(args, agreement));
+  const [responsePromises, usernamesPromises] = transpose(promiseMatrix);
+
+  // A new Promise that resolves to either
+  // 1. Username from the first Sign API call that includes the username
+  // 2. Empty array if none of the Sign API calls includes the username
+  // Promise is rejected if any of the Sign API calls fail before resolution
+  // If the username is found, all requests are aborted (it's okay to abort completed requests)
+  return new Promise((resolve, reject) => {
+    Promise.all(usernamesPromises.map(promise => {
+      return promise.then(agreementUsers => {
+        if (agreementUsers.map(u => u.toLowerCase()).includes(username)) {
+          resolve([username]);
+          responsePromises.forEach(p => p.abort());
+        }
+      });
+    })).then(_results => resolve([]), error => reject(error));
+  });
 }
 
-async function lookupSingleAgreement (args, agreement) {
+function lookupAndParseAgreement (args, agreement) {
+  // This isn't an ordinary Promise, it's actually a Request object that is also a Promise.
+  const responsePromise = lookupAgreement(args, agreement);
+  const usernamesPromise = responsePromise
+    .then(function (response) {
+      return parse(response.trim(), { columns: true });
+    })
+    .then(function (records) {
+      return Promise.resolve(extractUsernames(records));
+    });
+
+  return [responsePromise, usernamesPromise];
```

Comment:
Why does this function return two promises for each search request to the Sign API? I guess I'm wanting to understand the overall shape of your approach.